homeassistant/auth/__init__.py

```diff
@@ -46,6 +57,66 @@ def generate_secret(entropy: int = 32) -> str:
     return binascii.hexlify(os.urandom(entropy)).decode('ascii')
 
 
+class InvalidAuth(HomeAssistantError):
+    """Raised when we encounter invalid authentication."""
+
+
+class AuthModule:
+    """Provider of validation function."""
+
+    DEFAULT_TITLE = 'Unnamed auth module'
+
+    initialized = False
+
+    def __init__(self, hass, config):
+        """Initialize an auth module."""
+        self.hass = hass
+        self.config = config
+        _LOGGER.debug('auth module %s loaded.',
+                      self.type if self.id is None else "{}[{}]".format(
+                          self.type, self.id
+                      ))
+
+    @property
+    def id(self):  # pylint: disable=invalid-name
```

Comment:
I don't think that we need to copy the id approach of providers. There can be only one.

Reply:
They can have multiple instance, for example, I am planning to implement a mfa_module to use `hotp` algorithm, then send the code by one of configed HA notify service. So I will end up to have two module with same type, but have different config if I want to use email or SMS.